src/main/java/edu/mit/puzzle/cube/core/model/User.java

```diff
@@ -0,0 +1,26 @@
+package edu.mit.puzzle.cube.core.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.auto.value.AutoValue;
+
+import java.util.List;
+
+@AutoValue
+@JsonDeserialize(builder = AutoValue_User.Builder.class)
+public abstract class User {
+    @AutoValue.Builder
+    public static abstract class Builder {
+        @JsonProperty("username") public abstract Builder setUsername(String username);
+        @JsonProperty("permissions") public abstract Builder setPermissions(List<String> permissions);
+
+        public abstract User build();
+    }
+
+    public static Builder builder() {
+        return new AutoValue_User.Builder();
+    }
+
+    @JsonProperty("username") public abstract String getUsername();
+    @JsonProperty("permissions") public abstract List<String> getPermissions();
```

Comment:
Yeah, that's a good insight.

The only reason I put the list of permissions there is so that the
huntadmin client can decide which UI to show (e.g. don't show the "Start
Hunt" button to a user without hunt starting permissions). There are other
ways we could address this e.g. by returning a list of roles instead of a
list of permissions. Now that I think about it, I think the client deciding
based on roles instead of permissions would be better.

I'll revise the branch later tonight to return roles instead of
permissions, if you agree that sounds better.

On Mon, Jun 6, 2016 at 2:54 PM, TK Focht <notifications@github.com> wrote:

> In src/main/java/edu/mit/puzzle/cube/core/model/User.java
> <https://github.com/eastern-toys/Cube/pull/12#discussion_r65947239>:
>
> > +@JsonDeserialize(builder = AutoValue_User.Builder.class)
> > +public abstract class User {
> > +    @AutoValue.Builder
> > +    public static abstract class Builder {
> > +        @JsonProperty("username") public abstract Builder setUsername(String username);
> > +        @JsonProperty("permissions") public abstract Builder setPermissions(List<String> permissions);
> > +
> > +        public abstract User build();
> > +    }
> > +
> > +    public static Builder builder() {
> > +        return new AutoValue_User.Builder();
> > +    }
> > +
> > +    @JsonProperty("username") public abstract String getUsername();
> > +    @JsonProperty("permissions") public abstract List<String> getPermissions();
>
> This is going to be a problem, or at least a confusing inconsistency, if
> we assign permissions to roles rather than directly to users, which I think
> that we should due to having individual users at least on the game control
> team - and possibly standard and admin accounts for solving teams?
>
> Shiro doesn't appear to have a Subject.listPermissions() or equivalent,
> which I totally get because that is HARD, but it means that we'll have a
> hard time populating this field even on a Role object.
>
> Basically, I'm not sure what we would actually do with this in the model
> object, not sure we can make it actually accurate, and not sure it's in the
> right place.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eastern-toys/Cube/pull/12/files/1889a10fd3a13a670b3773b6e041e649b256509e#r65947239>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAgakNos1NNoXj4JAx9hY0YffZi_GNDhks5qJGzcgaJpZM4Iu5zY>
> .
>